Prepare for feature Forwarding CSR to upstream CA (External CA)

hovonako opened the pull request:
Hi Istio team, 

Based on the discussion about CAs Integration with IstioD integration approach ([RFC Link](https://docs.google.com/document/d/1KAw8-0FivdYQAcWMVTMl-JYvVCn2fpNTbu6Ya3y9d1E/edit#heading=h.1iv3g9pjzq73)). I'm introducing this pull request to preparing the istio/api for feature forwarding CSR to upstream CA within Istio(in-progressing), detailed in below diagram.
![Istio Forwarder](https://user-images.githubusercontent.com/49465776/87638139-79f9d400-c76d-11ea-8142-9499d6e4528f.jpg)

Feature pull request: [Link Pull Request](https://github.com/istio/istio/pull/25641)
This pull request includes:
  - istio.io/api/mesh/v1alpha1/config.proto: added variable **upstream_ca_address**
  - add comments to IstioCertificateRequest. In my opinion, Using gPRC's metadata to includes Pod's metadata will minimize impact on the current Istio codebase. In addition, these metadata are not closely related to the logic of CSR signing. So, it's better to be attached as gRPC's metadata.

Thanks,

istio-testing commented:
Hi @hovonako. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fapi).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

smawson commented:
Can you link to the RFC describing this feature? Or github issue where this was discussed with the security wg?

hovonako commented:
> Can you link to the RFC describing this feature? Or github issue where this was discussed with the security wg?

Hi @smawson , 
I've updated the RFC link and added the pull request for it.
Thanks,

costinm commented on mesh/v1alpha1/config.proto:
```diff
@@ -488,8 +488,14 @@ message MeshConfig {
+  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server 
+  // after the JWT is successfully checked (Authn & Authz)
+  // This feature is disabled by default (empty value)
+  string upstream_ca_address = 54;
```
Also: hide_from_docs until it is implemented and passes all the requirements ( testing, etc)

myidpt replied:
I agree with @costinm . We need at least:
ca_address
ca_tls_root_cert
client_authentication_method (k8s_jwt / tls_cert / OTP / etc.)

If you define it in mesh API, you also need to specify if this is applied to istiod or istio-agent (istio-agent also has ca_address). Given it's not trivial, so I'd suggest starting with istiod ENV rather than mesh API. We need to discuss whether we can configure this using DR. The details for the long term may warrant a design doc.

costinm commented on mesh/v1alpha1/config.proto:
```diff
@@ -488,8 +488,14 @@ message MeshConfig {
   reserved "termination_drain_duration";
 
+  // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto
+  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server 
```
I don't think 'instead of self-signing' is correct - Istiod is properly signing the CSR with its own key ( which can in turn be signed by another CA, in case if interemediate). 

Maybe 'If enabled, Istiod will delegate signing to another CA after authenticating the request and validating the SAN ?

costinm replied:
I would add: if this setting is enabled, Istiod will turn off its signing of Spiffee certificates. Changing this setting may disrupts traffic, see XXX for the process for migrating ( we need to document this - have both root CAs in the 'root CAs', make sure all workloads get the updated root bundle, use a new Istiod revision to test, etc)

rshriram replied:
can we please use the xds bootstrap metadata instead of messing around with grpc headers? We will not have to change the istio CA every time some vendor needs a new metadata from the proxy. What ever is in bootstrap is available to the upstream ca and its upto them to derive the values they want.

costinm commented on security/v1alpha1/ca.proto:
```diff
@@ -25,6 +25,7 @@ option go_package="istio.io/api/security/v1alpha1";
 // Note: the service implementation is REQUIRED to verify the authenticated caller is authorize to
 // all SANs in the CSR. The server side may overwrite any requested certificate field based on its
 // policies.
+// Some metadata of pod are included in gRPC's metadata: PodName, PodNamespace, PodIP, TrustDomain, ClusterID
```
I think ClusterID is currently included - can you dig the actual header we have ? 
The other ones are new:
- trustDomain is part of the spiffee spec and is in the CSR - do we really need to pass it again ? I don't mind, just checking.

PodName, PodNamespace are a bit more problematic - we generally avoid Istio APIs depending on K8S concepts.
Namespace will work - I don't know for the other ( WorkloadName ? ). Check what Telemetry is using, we want consistency between APIs.

Last one - PodIP should probably be WorkloadIPs, can be more than one - and more important any IP also needs a
Network ( multi-network feature requires this ).

We have discussed adding those elements in the certificate, as additional claims.

rshriram replied:
can we please use the xds bootstrap metadata instead of messing around with grpc headers? We will not have to change the istio CA every time some vendor needs a new metadata from the proxy. What ever is in bootstrap is available to the upstream ca and its upto them to derive the values they want.

hovonako replied:
hi @rshriram
Let's me dig the xds bootstrap metadata.
Thanks for the suggestion

rshriram replied:
There is no need to refer to the actual envoy xds metadata.. Importing that struct is going to cause a lot of protobuf pains.. All you need is a simple 

```
  // Opaque metadata provided by the XDS node to Istio.
  google.protobuf.Struct metadata = 3;
```

this struct is essentially a json.

hzxuzhonghu replied:
@costinm I am sure `clusterID` is included

hovonako replied:
Thanks @rshriram. 

@hzxuzhonghu yup, i saw. Currently, Cluster_id is included in gRPC Metadata. Therefore, I'm going to do the same way for WorkloadName, Namespace, IP, etc. :)))

smawson commented:
> > Can you link to the RFC describing this feature? Or github issue where this was discussed with the security wg?
> 
> Hi @smawson ,
> I've updated the RFC link and added the pull request for it.
> Thanks,

That doc has not yet been approved, so it is not clear if the approach described there has been agreed upon. Please get the doc signed off by the appropriate approvers (security and networking WG leads, possibly environment WG leads as well).

costinm commented:
The signing API has been merged into istio/api - I assume that means it is
at least directionally approved.
The RFCs have many options and tradeoffs, like many other docs - so not
clear what 'approving it' means,
some options are unlikely to happen.

As I suggested, using an env variable to do a prototype/experimental
feature first, before committing it to
an API would be ideal.


On Mon, Jul 20, 2020 at 10:12 AM Sven Mawson <notifications@github.com>
wrote:

> Can you link to the RFC describing this feature? Or github issue where
> this was discussed with the security wg?
>
> Hi @smawson <https://github.com/smawson> ,
> I've updated the RFC link and added the pull request for it.
> Thanks,
>
> That doc has not yet been approved, so it is not clear if the approach
> described there has been agreed upon. Please get the doc signed off by the
> appropriate approvers (security and networking WG leads, possibly
> environment WG leads as well).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#issuecomment-661202644>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2SXJVAB5D6NUXKV2RDR4R3GDANCNFSM4O3WWJSA>
> .
>

hovonako commented:
> Thanks for starting the API discussion early - let's not hurry with the API.
> Implementation can be done using the env variable, which is what we often use for configuring new/experimental stuff.
> 
> Would pilot use the external CA to get a cert for itself as well ? I assume yes..

Yup, pilot will use the external CA to get a Cert & Root Cert for itself.

rshriram commented on mesh/v1alpha1/config.proto:
```diff
@@ -488,8 +488,14 @@ message MeshConfig {
+  // after the JWT is successfully checked (Authn & Authz)
+  // This feature is disabled by default (empty value)
+  string upstream_ca_address = 54;
+
```
where are the TLS certs to talk to this CA?

hovonako replied:
@rshriram, Istio supports users plugging a root-cert.pem by using k8s secret. So, I will use it instead defining new one.

myidpt replied:
OK. You are suggesting we reuse the root-cert.pem for the Istiod-CA TLS connection purpose. I'm not totally against it, but will configuring the root cert for TLS connections separately be clearer? Note we need to also configure the client credentials (multiple types: cert / JWT / OTP / ...).

hovonako replied:
@myidpt ,
For integration between external CA and Istiod, I think we only need TLS connection. External CA should deploy a GRPC Serivce Gateway inside K8S (or maybe sidecar of istiod), will be implemented based on signing API and will responsible to forward the Workload CSR to External CA with it's credentials. It's more flexible than support configuring client credentials inside Istiod.

GRPC Serivce Gateway may verify k8s's JWT itself

rshriram replied:
where are the CA certs to verify the TLS connection, from istiod side?

costinm commented:
The API reflects what we currently have - that was the consensus, it has
been a stable API for a long
time.

I agree that messing with headers was a bad idea - and since the interface
is in effect sending a binary
blob and gets back a binary blob ( csr -> cert ), it might have been better
to use a simple POST instead
of gRPC, which is closer to ACME.

I also agree that we could add XDS metadata - and if we do this, maybe just
use the DiscoveryRequest
format - which makes it very close to SDS with an additional payload. There
are issues tracking this
in Envoy, and it would be great to move ahead with them.

But all this is discussion about a new API replacing the current one - i.e.
a new API. And I think
we also have agreement on doing this - with longer discussions on which one
to use (ACME,
extended XDS/SDS, etc ).

My request is to not mix 'new API' into the 'super stable API we had for 2
years'.




On Mon, Jul 20, 2020 at 6:19 PM Shriram Rajagopalan <
notifications@github.com> wrote:

> *@rshriram* commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/1531#discussion_r457779570>:
>
> > @@ -488,8 +488,14 @@ message MeshConfig {
>    reserved 53;
>    reserved "termination_drain_duration";
>
> +  // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto
> +  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server
>
> can we please use the xds bootstrap metadata instead of messing around
> with grpc headers? We will not have to change the istio CA every time some
> vendor needs a new metadata from the proxy. What ever is in bootstrap is
> available to the upstream ca and its upto them to derive the values they
> want.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#discussion_r457779570>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2X7YW7PUCYBMKZQMLLR4TUL3ANCNFSM4O3WWJSA>
> .
>

costinm commented:
The TLS cert to talk with the CA is in the existing config we use for
setting the root certs for Istio.
Let's keep Istiod consistent with the agent and not invent a new mechanism.

If we really need to fine tune the config - we already have a
DestinationRule, which Istiod
can use either directly or using a sidecar, no need to add yet another API
duplicating DR.

On Tue, Jul 21, 2020 at 10:33 AM Shriram Rajagopalan <
notifications@github.com> wrote:

> *@rshriram* commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/1531#discussion_r458270343>:
>
> > @@ -488,8 +488,14 @@ message MeshConfig {
>    reserved 53;
>    reserved "termination_drain_duration";
>
> +  // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto
> +  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server
> +  // after the JWT is successfully checked (Authn & Authz)
> +  // This feature is disabled by default (empty value)
> +  string upstream_ca_address = 54;
> +
>
> where are the TLS certs to talk to this CA?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#pullrequestreview-452681024>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2VDEQLQ77U7RRNJASDR4XGNRANCNFSM4O3WWJSA>
> .
>

costinm commented:
If I can ask a favor: please add a test with 2 root CAs in that file, so we
can also migrate from citadel to the external CA.

Bonus points if new root is EC256 - we need that migration too.

On Tue, Jul 21, 2020, 17:18 Khoa Hoàng <notifications@github.com> wrote:

> *@hovonako* commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/1531#discussion_r458461337>:
>
> > @@ -488,8 +488,14 @@ message MeshConfig {
>    reserved 53;
>    reserved "termination_drain_duration";
>
> +  // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto
> +  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server
> +  // after the JWT is successfully checked (Authn & Authz)
> +  // This feature is disabled by default (empty value)
> +  string upstream_ca_address = 54;
> +
>
> @rshriram <https://github.com/rshriram>, Istio supports users plugging a
> root-cert.pem by using k8s secret. So, I will use it instead defining new
> one.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#discussion_r458461337>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2WZLCVGNXMYIWJQKNTR4YV35ANCNFSM4O3WWJSA>
> .
>

hzxuzhonghu commented on mesh/v1alpha1/config.proto:
```diff
@@ -488,8 +488,14 @@ message MeshConfig {
+  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server 
+  // after the JWT is successfully checked (Authn & Authz)
+  // This feature is disabled by default (empty value)
+  string upstream_ca_address = 54;
```
I have a doubt: why not just make it as a startup param of istiod? I can not see the necessity  to put it in mesh configuration

hovonako replied:
thank @hzxuzhonghu ,
I'll consider your suggestion. I will first use env variable for the prototype before this pull request.

myidpt replied:
Same as above, I think starting with istiod ENV is a better choice.

hovonako commented:
> If I can ask a favor: please add a test with 2 root CAs in that file, so we can also migrate from citadel to the external CA. Bonus points if new root is EC256 - we need that migration too.
> […](#)
> On Tue, Jul 21, 2020, 17:18 Khoa Hoàng ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In mesh/v1alpha1/config.proto <[#1531 (comment)](https://github.com/istio/api/pull/1531#discussion_r458461337)>: > @@ -488,8 +488,14 @@ message MeshConfig { reserved 53; reserved "termination_drain_duration"; + // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto + // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server + // after the JWT is successfully checked (Authn & Authz) + // This feature is disabled by default (empty value) + string upstream_ca_address = 54; + @rshriram <https://github.com/rshriram>, Istio supports users plugging a root-cert.pem by using k8s secret. So, I will use it instead defining new one. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#1531 (comment)](https://github.com/istio/api/pull/1531#discussion_r458461337)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAUR2WZLCVGNXMYIWJQKNTR4YV35ANCNFSM4O3WWJSA> .

well-noted, @costinm .

costinm commented:
I think the important question is if we want to keep abusing gRPC metadata
- or add the fields to the proto.
We can extend the proto. And going back to Shriram suggestion - it wouldn't
be bad to reuse the node
 metadata struct that we already use for XDS, i.e. same struct and field
names, for consistency.

Do you really need PodIP and PodName ? As I mentioned, I want to avoid
k8s-specific names.

On Wed, Jul 22, 2020 at 6:17 AM Khoa Hoàng <notifications@github.com> wrote:

> *@hovonako* commented on this pull request.
> ------------------------------
>
> In security/v1alpha1/ca.proto
> <https://github.com/istio/api/pull/1531#discussion_r458783423>:
>
> > @@ -25,6 +25,7 @@ option go_package="istio.io/api/security/v1alpha1";
>  // Note: the service implementation is REQUIRED to verify the authenticated caller is authorize to
>  // all SANs in the CSR. The server side may overwrite any requested certificate field based on its
>  // policies.
> +// Some metadata of pod are included in gRPC's metadata: PodName, PodNamespace, PodIP, TrustDomain, ClusterID
>
> Thanks @rshriram <https://github.com/rshriram>.
>
> @hzxuzhonghu <https://github.com/hzxuzhonghu> yup, i saw. Currently,
> Cluster_id is included in gRPC Metadata. Therefore, I'm going to do the
> same way for WorkloadName, Namespace, IP, etc. :)))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#discussion_r458783423>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2WN3F7S5GL4YBKU7ADR43RHJANCNFSM4O3WWJSA>
> .
>

costinm commented:
"mesh configuration" is really the Istiod config. We no longer have a
global mesh config - each revision has its own
config.

We are trying to keep the 'stable/supported/long term' options of Istiod in
MeshConfig - and use env variables for
experimental/new features. In 1.8 we'll likely move more things from
global.yaml and values.yaml into MeshConfig.

I agree that this is not yet ready for API, but in 1-2 releases I think
ca_address should be added.

In ProxyConfig we'll also define a ca_addr - since the CA_ADDR env variable
has been stable and is clearly needed.

On Wed, Jul 22, 2020 at 3:19 AM Zhonghu Xu <notifications@github.com> wrote:

> *@hzxuzhonghu* commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/1531#discussion_r458689288>:
>
> > @@ -488,8 +488,14 @@ message MeshConfig {
>    reserved 53;
>    reserved "termination_drain_duration";
>
> +  // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto
> +  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server
> +  // after the JWT is successfully checked (Authn & Authz)
> +  // This feature is disabled by default (empty value)
> +  string upstream_ca_address = 54;
>
> I have a doubt: why not just make it as a startup param of istiod? I can
> not see the necessity to put it in mesh configuration
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#pullrequestreview-453185308>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2XSXGTNOFVHYSAVPY3R424K3ANCNFSM4O3WWJSA>
> .
>

hovonako commented:
> I think the important question is if we want to keep abusing gRPC metadata - or add the fields to the proto. We can extend the proto. And going back to Shriram suggestion - it wouldn't be bad to reuse the node metadata struct that we already use for XDS, i.e. same struct and field names, for consistency. Do you really need PodIP and PodName ? As I mentioned, I want to avoid k8s-specific names.

Hi @costinm,
I will use the name: WorkloadName , Workload IP instead of k8s-specific names. Our customers need these fields to monitor ,analyze and reports.
I also update the code based on @rshriram's suggestion (using google.protobuf.Struct)

Thanks again,

myidpt commented:
@smawson This implementation is based on the existing CSR API and for this workflow, we have got approvals from the security WG (we have already published the CSR API to istio/api).
For the doc https://docs.google.com/document/d/1KAw8-0FivdYQAcWMVTMl-JYvVCn2fpNTbu6Ya3y9d1E/edit#heading=h.12ce7636lrgb we want to make sure all reasonable integration approaches are allowed, as long as it doesn't complicate Istio implementation (to @costinm 's concern). We have approvals from the security WG, I will push for approval from networking and env WGs.

costinm commented:
We should clearly document the current logic for determining the roots -
for both CA and XDS endpoints.

It is based on the default paths - with some env to override. The paths and
the k8s secrets that get mounted are internal right now - but have been
stable enough and are worth making part of the contract.

In this case, the admin is responsible to create the Secret with the list
of 'mesh roots' - including the external CA roots ( can have both rsa and
ec, plus keys for rotation/migration).
Please don't invent a new one !



On Sun, Jul 26, 2020, 05:36 Shriram Rajagopalan <notifications@github.com>
wrote:

> *@rshriram* commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/1531#discussion_r460522291>:
>
> > @@ -488,8 +488,14 @@ message MeshConfig {
>    reserved 53;
>    reserved "termination_drain_duration";
>
> +  // The address of upstream CA, which implements istio.io/api/security/v1alpha1/ca.proto
> +  // If enabled, Instead of self-signing IstioAgent's CSR, Istio Discovery will forward the CSR to upstream CA server
> +  // after the JWT is successfully checked (Authn & Authz)
> +  // This feature is disabled by default (empty value)
> +  string upstream_ca_address = 54;
> +
>
> where are the CA certs to verify the TLS connection, from istiod side?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/1531#discussion_r460522291>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2TJVWWQFPG4W3YHZRLR5QPMFANCNFSM4O3WWJSA>
> .
>

istio-testing commented:
@hovonako: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
